ap/inference/datacard.py

```diff
@@ -0,0 +1,550 @@
+        # expected rates
+        proc_names, s_names, b_names = [], [], []
+        flat_rates = OrderedDict()
+        for cat_name, _rates in rates.items():
+            for proc_name, rate in _rates.items():
+                if proc_name == "data":
+                    continue
+
+                # devide into signal and backgrounds
+                if proc_name not in proc_names:
+                    proc_obj = self.inference_model_inst.get_process(proc_name, category=cat_name)
+                    (s_names if proc_obj.signal else b_names).append(proc_name)
+
+                # fill flat rates
+                flat_rates[(cat_name, proc_name)] = rate
+
+        blocks.rates = [
+            ("bin", [cat_name for cat_name, _ in flat_rates]),
+            ("process", [proc_name for _, proc_name in flat_rates]),
+            ("process", [
+                (-s_names.index(proc_name) if proc_name in s_names else b_names.index(proc_name) + 1)
+                for _, proc_name in flat_rates
+            ]),
+            ("rate", [round(rate, self.rate_precision) for rate in flat_rates.values()]),
+        ]
+        separators.add("rates")
+
+        # tabular-style parameters
+        blocks.tabular_parameters = []
+        rnd = lambda f: round(f, self.parameter_precision)
+        for param_name in self.inference_model_inst.get_parameters(flat=True):
+            param_obj = None
+            effects = []
+            for cat_name, proc_name in flat_rates:
+                _param_obj = self.inference_model_inst.get_parameter(
+                    param_name,
+                    category=cat_name,
+                    process=proc_name,
+                    silent=True,
+                )
+
+                # skip line-style parameters as they are handled separately below
+                if _param_obj and _param_obj.type == ParameterType.rate_unconstrained:
+                    continue
+
+                # empty effect
+                if _param_obj is None:
+                    effects.append("-")
+                    continue
+
+                # compare with previous param_obj
+                if param_obj is None:
+                    param_obj = _param_obj
+                elif _param_obj.type != param_obj.type:
+                    raise ValueError(
+                        f"misconfigured parameter '{param_name}' with type '{_param_obj.type}' "
+                        f"that was previously seen with incompatible type '{param_obj.type}'")
+
+                # get the effect
+                effect = _param_obj.effect
+
+                # update and transform effects
+                if _param_obj.type.is_rate:
+                    # obtain from shape effects when requested
+                    if _param_obj.transformations.any_from_shape:
+                        effect = shape_effects[cat_name][proc_name][param_name]
+
+                    # apply transformations one by one
+                    for trafo in _param_obj.transformations:
```

Comment:
Yeah, it already felt like code duplication while writing. I think I would rather have that in the datacard writer, since that's the one that also does the other transformations.

Reply:
Ah you mean the InferenceModel class can't do the transformations because it doesn't have the templates, right? It's also fine by me if the functions are in the DatacardWriter, but I think it would be good to centralize this from the start. Otherwise, I feel like this could generate confusion rather quickly

Reply:
True. I removed them